plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Core.hs

```diff
@@ -1,7 +1,27 @@
-module UntypedPlutusCore.Core
-    ( module Export
-    ) where
+{-# LANGUAGE LambdaCase #-}
+
+module UntypedPlutusCore.Core (
+  module Export,
+  splitApplication,
+  mkApplication,
+) where
 
 import UntypedPlutusCore.Core.Instance as Export
 import UntypedPlutusCore.Core.Plated as Export
 import UntypedPlutusCore.Core.Type as Export
+
+import Data.List (foldl')
+
+type Arg name uni fun a = (Term name uni fun a, a)
+
+-- | Strip off arguments
+splitApplication :: Term name uni fun a -> (Term name uni fun a, [Arg name uni fun a])
+splitApplication = go []
+  where
+    go acc = \case
+      Apply ann fun arg -> go ((arg, ann) : acc) fun
+      t                 -> (t, acc)
+
+-- | Make an Apply node from the given function and arguments.
+mkApplication :: Term name uni fun a -> [Arg name uni fun a] -> Term name uni fun a
+mkApplication = foldl' $ \acc (arg, ann) -> Apply ann acc arg
```

Comment:
This is like

```haskell
mkIterApp
    :: TermLike term tyname name uni fun
    => ann
    -> term ann -- ^ @f@
    -> [term ann] -- ^@[ x0 ... xn ]@
    -> term ann -- ^ @[f x0 ... xn ]@
mkIterApp ann = foldl' (apply ann)
```

except doesn't share a single annotation for application nodes. Maybe we should keep this function in `MkPlc` too? And maybe we should generally do something about annotations being shared/not shared. Like provide some kind of convenient API and allow the user to choose or something idk.

Reply:
The PIR counterpart is `PlutusIR.Contexts.fillAppContext`. Perhaps they should all use the `TermLike` API in `MkPLC`?

I don't think sharing annotation makes sense. `splitApplication` and `mkApplication` should round-trip.

Reply:
> I don't think sharing annotation makes sense.

I think it does sometimes, but I agree that not in this case.